kotlin-grpc/src/test/kotlin/com/linecorp/armeria/server/grpc/CoroutineServerInterceptorTest.kt

```diff
@@ -0,0 +1,100 @@
+import com.linecorp.armeria.testing.junit5.server.ServerExtension
+import io.grpc.Metadata
+import io.grpc.ServerCall
+import io.grpc.ServerCallHandler
+import io.grpc.Status
+import kotlinx.coroutines.future.await
+import org.junit.jupiter.api.extension.RegisterExtension
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+
+internal class CoroutineServerInterceptorTest {
+    private class AuthInterceptor : CoroutineServerInterceptor {
+        private val authorizer = Authorizer { ctx: ServiceRequestContext, metadata: Metadata ->
+            val future = CompletableFuture<Boolean>()
+            ctx.eventLoop().schedule({
+                if (ctx.request().headers().contains("Authorization", "Bearer token-1234")) {
+                    future.complete(true)
+                } else {
+                    future.complete(false)
+                }
+            }, 100, TimeUnit.MILLISECONDS)
+            return@Authorizer future
+        }
+
+        override suspend fun <ReqT, RespT> suspendedInterceptCall(
+            call: ServerCall<ReqT, RespT>,
+            headers: Metadata,
+            next: ServerCallHandler<ReqT, RespT>
+        ): ServerCall.Listener<ReqT> {
+            val result = authorizer.authorize(ServiceRequestContext.current(), headers).await()
+            if (result) {
+                return next.startCall(call, headers)
+            } else {
+                throw AnticipatedException("Invalid access")
+            }
+        }
+    }
+
+    companion object {
+        @RegisterExtension
+        val server: ServerExtension = object : ServerExtension() {
+            override fun configure(sb: ServerBuilder) {
+                val statusFunction = GrpcStatusFunction { ctx: RequestContext?, throwable: Throwable, metadata: Metadata? ->
+                    if (throwable is AnticipatedException && "Invalid access" == throwable.message) {
+                        return@GrpcStatusFunction Status.UNAUTHENTICATED
+                    }
+                    null
+                }
+                val authInterceptor = AuthInterceptor()
+                sb.serviceUnder(
+                    "/non-blocking",
+                    GrpcService.builder()
+                        .exceptionMapping(statusFunction)
+                        .intercept(authInterceptor)
+                        .build()
+                )
+                sb.serviceUnder(
+                    "/blocking",
+                    GrpcService.builder()
+                        .exceptionMapping(statusFunction)
+                        .intercept(authInterceptor)
+                        .useBlockingTaskExecutor(true)
+                        .build()
+                )
+            }
+        }
+
+        object TestServiceImpl {
+            fun todo() {
+                TODO("If we need same test codes like AsyncServerInterceptorTest, need TestServiceImpl in grpc module's test dir")
+            }
+        }
```

Comment:
If I understood correctly, you can just copy the `.proto` file into `kotlin-grpc/src/test/proto` to generate `TestServiceImplBase` then extend it here. Some level of duplication is OK I guess.

Reply:
Thank you for your comments.
We'll copy the proto file and write a test case for the new Interceptor 🙂